Upgrade to 1.x standards (#10)
inviqa/harness-base-node: README.md

> # Harness Base PHP for [Workspace]

Reviewer: ```suggestion
# Harness Base Node for [Workspace]
```

A [Workspace] harness is a way to ship files to a project without being part of the project.

In this repository are a set of harnesses that have been created for the PHP language, reducing the maintenance overhead

> of the individual harnesses greatly. This is due to the "base" PHP harness being used as a common set of templates.

Reviewer: ```suggestion
In this repository are a set of harnesses that have been created for the NodeJS language, reducing the maintenance overhead
of the individual harnesses greatly. This is due to the "base" NodeJS harness being used as a common set of templates.
```

8. Compare the overrides for attributes in the project's `workspace.yml` to the harness's `harness.yml` and `harness/attributes/*.yml`.
    1. Port over any additional build, init or migrate steps.
    2. Remove any attribute overrides from workspace.yml that are now the same as the harness.
9. Test with `ws harness update existing`
10. Open a pull request with the project and ensure CI checks (such as Jenkins) pass.
11. Note on the pull request that to apply the changes, the following needs to be done by the project team:
```markdown
To keep your existing database:
`ws harness update existing`

To do a fresh installation:
`ws harness update fresh`
```
12. Ask for someone else to test the pull request.
13. Once the pull request has been merged to the default branch of the project, remind the project team to apply the changes with the `ws harness update existing` or `ws harness update fresh` commands.

## Helm charts

Each harness deploys:
* A "console" pod for running one-off commands
* A NGINX/PHP-FPM "webapp" pod for running the php-based application and serving web requests
* A service to route to the "webapp" pods

> * An ingress definition to route via the "webapp" service

Reviewer: Needs updating to reflect what we do here

## Testing

The final harness version for each of the frameworks is put together by the [build script](./build) into a "dist"
folder. This is used for testing in Jenkins.

### Static mode

1. Run `./build`
2. Create and change to a tests directory
```bash
mkdir tests
cd tests
ws create <framework>-test inviqa/<framework> --no-install
```
3. Copy the built version of harness-base-php's ``./dist/<framework>`` dir to `<framework>-test/.my127ws` directory
```bash
cp -pR ../dist/harness-<framework>/ <framework>-test/.my127ws/
```
4. Change directory to `<framework>-test`
```bash
cd `<framework>-test`
```
5. Update the override file in `<framework>-test`:
```bash
echo "attribute('mutagen'): no" >> workspace.override.yml
```
6. Run in `pipeline` mode to activate `static` mode
```bash
MY127WS_ENV=pipeline ws install
```

## Deployment

### Deploy Script

The [deploy script](./deploy) does a similar thing but the end result is output to `src/<framework>`, where it is then
committed as a publish commit to a temporary build branch.

A "subtree-split" is then performed which outputs a directory for each folder into a "publish" folder, where it is then

> force pushed to the individual harness repositories' `1.1.x` branch.

Reviewer: 1.1.x and references to 1.1.1/1.2.0 will need to be updated based on the next release version number

  ```bash
  git checkout -b 1.2.x
  ```
2. Adjust references from 1.1.x to 1.2.x:
  ```bash
  grep -FR '1.2.x' . | grep -v dist/
  grep -FR '1.1.x' . | grep -v dist/
  # Edit resulting files
  ```
3. Adjust references in this file from 1.1.1 to 1.2.0:
  ```bash
  grep -FR '1.2.0' README.md
  grep -FR '1.1.1' README.md
  # Edit resulting files
  ```
4. Commit the resulting files and push:
  ```bash
  git add -p
  git commit
  git push origin -u HEAD
  ```
5. Change the default branch in GitHub settings and re-target any open PRs against the new default branch.
6. Run a deployment of the 1.1.x branch so the branches now exist in the downstream repositories:
  ```bash
  ./deploy
  ```
7. Adjust each downstream repository's default branch so people can easily follow the README

[Workspace]: https://github.com/my127/workspace
[GitHub Changelog Generator]: https://github.com/github-changelog-generator/github-changelog-generator
[inviqa/harness-node]: https://github.com/inviqa/harness-node
[inviqa/harness-node-spa]: https://github.com/inviqa/harness-node-spa
[inviqa/harness-viper]: https://github.com/inviqa/harness-viper
[my127/my127.io]: https://github.com/my127/my127.io

> [version specific upgrade instructions]: https://github.com/inviqa/harness-base-php/blob/1.1.x/UPGRADE.md

Reviewer: ```suggestion
[version specific upgrade instructions]: https://github.com/inviqa/harness-base-node/blob/0.2.x/UPGRADE.md
```
